virtio-devices, net_util, vhost_user_net: Retry writing to TAP

rbradford opened the pull request:
If writing to the TAP returns EAGAIN then listen for the TAP to be
writable. When the TAP becomes writable attempt to process the TX queue
again.

Fixes: #2807

LeeSmet commented on net_util/src/queue_pair.rs:
```diff
@@ -271,8 +273,35 @@ impl NetQueuePair {
+            .map_err(NetQueuePairError::UnregisterListener)?;
+            self.tx_tap_listening = true;
+            info!("Writing to TAP returned EAGAIN. Listening for TAP to become writable.");
+        } else if self.tx_tap_listening {
```
I might be wrong here, but I think this should also check to make sure there was no EAGAIN. Right now, if a listener is set up, it will be removed on the next call of this method. In practice, from the logs I saw (prior to running this patch), this function might run multiple times in quick succession where each invocation hits EAGAIN. In that scenario, the second call would remove the listener again. Likewise, afaik there is no guarantee that a tap becoming writeable again is able to send all the pending data, yet the listener would be removed as well.

rbradford replied:
@LeeSmet I actually tried to handle the successive EAGAIN hence the `else if`. If `process_desc_chain` encounters EAGAIN then it returns true (`tx_tap_retry`).

If we get told we need to retry and no listener is set then add the listener (`if tx_tap_retry && !self.tx_tap_listening`) if we don't need to retry and a listener is set ` } else if self.tx_tap_listening {` remove the listener.

So if we have to retry and we already have the listener then none of the above code will be run and the listener retained.

Please be aware that the `else if` is not re-evalulated. So this is equivalent to.

`if tx_tap_retry && !self.tx_tap_listening { .. } else { if self.tx_tap_listening { .. }  }` however clippy will complain about this style.

rbradford replied:
Oops. Scratch that. I see I was missing an extra clause on the `} else if {`

rbradford replied:
Thanks! :-)

LeeSmet commented:
While trying this patch:

```
cloud-hypervisor: 502.010151348s: <_net2_qp0> WARN:net_util/src/queue_pair.rs:99 -- net: tx: (recoverable) failed writing to tap: Resource temporarily unavailable (os error 11)
cloud-hypervisor: 502.010369035s: <_net2_qp0> ERROR:virtio-devices/src/net.rs:273 -- Error processing TX queue: NetQueuePair(UnregisterListener(Os { code: 17, kind: AlreadyExists, message: "File exists" }))
```

After which the affected interface is dead

rbradford commented:
> While trying this patch:
> 
> ```
> cloud-hypervisor: 502.010151348s: <_net2_qp0> WARN:net_util/src/queue_pair.rs:99 -- net: tx: (recoverable) failed writing to tap: Resource temporarily unavailable (os error 11)
> cloud-hypervisor: 502.010369035s: <_net2_qp0> ERROR:virtio-devices/src/net.rs:273 -- Error processing TX queue: NetQueuePair(UnregisterListener(Os { code: 17, kind: AlreadyExists, message: "File exists" }))
> ```
> 
> After which the affected interface is dead

Is that with the fixed version?

LeeSmet commented:
> > While trying this patch:
> > ```
> > cloud-hypervisor: 502.010151348s: <_net2_qp0> WARN:net_util/src/queue_pair.rs:99 -- net: tx: (recoverable) failed writing to tap: Resource temporarily unavailable (os error 11)
> > cloud-hypervisor: 502.010369035s: <_net2_qp0> ERROR:virtio-devices/src/net.rs:273 -- Error processing TX queue: NetQueuePair(UnregisterListener(Os { code: 17, kind: AlreadyExists, message: "File exists" }))
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > After which the affected interface is dead
> 
> Is that with the fixed (as in the version with the `else if` fixed) version?

Yes, rebuild after you pushed that

rbradford commented:
> > > While trying this patch:
> > > ```
> > > cloud-hypervisor: 502.010151348s: <_net2_qp0> WARN:net_util/src/queue_pair.rs:99 -- net: tx: (recoverable) failed writing to tap: Resource temporarily unavailable (os error 11)
> > > cloud-hypervisor: 502.010369035s: <_net2_qp0> ERROR:virtio-devices/src/net.rs:273 -- Error processing TX queue: NetQueuePair(UnregisterListener(Os { code: 17, kind: AlreadyExists, message: "File exists" }))
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > After which the affected interface is dead
> > 
> > 
> > Is that with the fixed (as in the version with the `else if` fixed) version?
> 
> Yes, rebuild after you pushed that

I can see the problem is that each FD in the on epoll must be unique.

rbradford commented:
> > > > While trying this patch:
> > > > ```
> > > > cloud-hypervisor: 502.010151348s: <_net2_qp0> WARN:net_util/src/queue_pair.rs:99 -- net: tx: (recoverable) failed writing to tap: Resource temporarily unavailable (os error 11)
> > > > cloud-hypervisor: 502.010369035s: <_net2_qp0> ERROR:virtio-devices/src/net.rs:273 -- Error processing TX queue: NetQueuePair(UnregisterListener(Os { code: 17, kind: AlreadyExists, message: "File exists" }))
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > After which the affected interface is dead
> > > 
> > > 
> > > Is that with the fixed (as in the version with the `else if` fixed) version?
> > 
> > 
> > Yes, rebuild after you pushed that
> 
> I can see the problem is that each FD in the on epoll must be unique.

I will need to make some significant changes to handle this but I have a way to reproduce the problem by faking EAGAIN based on a random value. I'll get back in touch when I have a solution that works with that.

rbradford commented:
@LeeSmet Thank you for your patience. I've updated the PR with a version that will avoid the epoll problem that you were seeing. Since I cannot reproduce your problem directly (checking the kernel code -EAGAIN seems to occur when the skb cannot be allocated) I modified the code to 25% of the time behave as if it failed to write with -EAGAIN, with this high probability it also showed up double -EAGAINs.

Please can you test this code firstly to check that this does not introduce any more failures when deployed in practice and also whether this in fact helps the underlying problem you were seeing.

rbradford commented:
Note to reviewers: The solution of using a second TAP fd is a little bit janky but it is one of the recommended solutions to watching the same file in the epoll man page if you want to have different events triggered. The alternative is to manage the watch on the fd using `EPOLL_CTL_MOD` to add/remove the different events and then have just one event e.g. `TAP_EVENT` and inside the event loop check what type of event it was (`EPOLLOUT` / `EPOLLIN`). However we currently manage the RX event in many circumstances for rate limiting and also if the RX queue is saturated. Trying to handle that and modifying the event on the loop for the different event type without some careful refactoring would be very fragile.

LeeSmet commented:
Running this patch on the node where I got the issue (not sure why it happens either, the nodes network is not saturated and neither is the switch its connected to). Seems to be working great :).

One small comment though, is it possible to downgrade the WARN log when encountering EAGAIN? Considering this is a supported scenario, and after some time (about 5 1 minute bursts of traffic) my log file really fills (CH started without any `-v` flags):

```
~ # wc -l /var/cache/modules/vmd/logs/45679-1
101489 /var/cache/modules/vmd/logs/45679-1
~ # cat /var/cache/modules/vmd/logs/45679-1 | grep "failed writing to tap" | wc -l
101489
```

Anyway, thanks for the time and effort on this.

rbradford commented:
> Running this patch on the node where I got the issue (not sure why it happens either, the nodes network is not saturated and neither is the switch its connected to). Seems to be working great :). In the (limited) time I've been testing it, I did not encounter any new errors

Great to hear! 

> One small comment though, is it possible to downgrade the WARN log when encountering EAGAIN? Considering this is a supported scenario, and after some time (about 5 1 minute bursts of traffic) my log file really fills (CH started without any `-v` flags):

Sure, i've dropped it which is consistent with the read case and we've now got `info!()` level messages around this code anyway.

sboeuf commented on net_util/src/queue_pair.rs:
```diff
@@ -93,12 +93,12 @@ impl TxVirtio {
                         warn!("net: tx: (recoverable) failed writing to tap: {}", e);
-                        break;
+                        queue.go_to_previous_position();
+                        return Ok(true);
```
This looks good but you could have reused the same format that is used for RxVirtio, by using `break` and `exhausted_descs` mutable variable.

likebreath reviewed (approved):
LGTM. Now the TX is fairly similar to the RX handling with the the `tx_tap_event` added. 

I will need to revisit the `rate_limiter` on the TX part, as at least its handling on the `EAGAIN` is not correct (the `rate_limiter` should replenish as the tokens are not consumed). I expect the TX `rate_limiter` should be very similar to the one for RX now. I will open an issue and revisit it later.
